File: src/tests/concurrency_test.js
// Comprehensive Concurrency Tests for Kode Runtime
// Testing Go-style goroutines, channels, and preemptive scheduling

console.log("=== Kode Concurrency Tests ===");

// Test 1: Basic Task Spawning
console.log("\n1. Testing basic task spawning:");
kode("console.log('Task 1 executed')");
kode("console.log('Task 2 executed')");
kode("console.log('Task 3 executed')");

// Test 2: Channel Communication
console.log("\n2. Testing channel communication:");
createChannel("test-channel", 2);  // Buffered channel with capacity 2

// Producer tasks
kode("sendToChannel('test-channel', 'Hello from producer')");
kode("sendToChannel('test-channel', 'Second message')");

// Consumer tasks (receive prints the value)
kode("receiveFromChannel('test-channel')");
kode("receiveFromChannel('test-channel')");

// Test 3: Cooperative Yielding
console.log("\n3. Testing cooperative yielding:");
kode("console.log('Before yield')");
kode("yield()\n");
kode("console.log('After yield')");

// Test 4: Timeout and Cancellation
console.log("\n4. Testing timeout functionality:");
withTimeout(100, "console.log('This should complete within 100ms')");

// Test 5: Heavy Computation (for preemption testing)
console.log("\n5. Testing preemptive scheduling (simplified demonstration):");
kode("console.log('Heavy task 1 simulated')");
kode("console.log('Heavy task 2 simulated')");
kode("console.log('Light task executed between heavy tasks')");

// Test 6: Multiple Channels
console.log("\n6. Testing multiple channels:");
createChannel("numbers", 0);   // Unbuffered channel
createChannel("strings", 3);   // Buffered channel

kode("sendToChannel('numbers', '42')");
kode("sendToChannel('strings', 'hello')");
kode("sendToChannel('strings', 'world')");

kode("receiveFromChannel('numbers')");
kode("receiveFromChannel('strings')");
kode("receiveFromChannel('strings')");

// Test 7: File System + Concurrency
console.log("\n7. Testing concurrent file operations:");
kode("fs.writeFile('concurrent-test-1.txt', 'Content from task 1')");
kode("fs.writeFile('concurrent-test-2.txt', 'Content from task 2')");
kode("fs.readFile('concurrent-test-1.txt')");

// Test 8: Mixed Sync/Async Operations
console.log("\n8. Testing mixed sync/async operations:");
kode("console.log('Sync operation')");
kode("fs.readFileSync('index.js')");
kode("console.log('Sync done')");
kode("console.log('Async start')");
kode("fs.readFile('index.js')");
kode("console.log('Async started')");

console.log("\nAll concurrency tests queued!");
console.log("Results will appear as tasks execute concurrently...");
console.log("Watch for preemption, context switching, and channel operations!");

File: src/tests/concurrency_test.cc
#include "../concurrency/task.h"
#include <iostream>
#include <chrono>
#include <thread>
#include <vector>
#include <atomic>

// Test the low-level concurrency implementation
void test_basic_task_creation() {
    std::cout << "\n=== Test 1: Basic Task Creation ===" << std::endl;
    
    ConcurrencyRuntime runtime;
    runtime.initialize(2);  // 2 worker threads
    
    std::atomic<int> counter{0};
    
    // Spawn multiple tasks
    for (int i = 0; i < 5; i++) {
        runtime.kode([i, &counter]() {
            std::cout << "[Task " << i << "] Starting execution" << std::endl;
            std::this_thread::sleep_for(std::chrono::milliseconds(100));
            counter.fetch_add(1);
            std::cout << "[Task " << i << "] Completed (counter: " << counter.load() << ")" << std::endl;
        });
    }
    
    runtime.join_all();
    std::cout << "Final counter value: " << counter.load() << std::endl;
    runtime.shutdown();
}

void test_channel_communication() {
    std::cout << "\n=== Test 2: Channel Communication ===" << std::endl;
    
    ConcurrencyRuntime runtime;
    runtime.initialize(2);
    
    auto channel = runtime.make_channel<std::string>(2);  // Buffered channel
    
    // Producer task
    runtime.kode([channel]() {
        std::cout << "[Producer] Sending messages..." << std::endl;
        channel->send("Hello");
        channel->send("World");
        channel->send("From");
        channel->send("Channels");
        channel->close();
        std::cout << "[Producer] All messages sent and channel closed" << std::endl;
    });
    
    // Consumer task
    runtime.kode([channel]() {
        std::cout << "[Consumer] Receiving messages..." << std::endl;
        std::string msg;
        while (channel->receive(msg)) {
            std::cout << "[Consumer] Received: " << msg << std::endl;
            std::this_thread::sleep_for(std::chrono::milliseconds(50));
        }
        std::cout << "[Consumer] Channel closed, stopping" << std::endl;
    });
    
    runtime.join_all();
    runtime.shutdown();
}

void test_preemptive_scheduling() {
    std::cout << "\n=== Test 3: Cooperative Scheduling Behavior ===" << std::endl;
    
    ConcurrencyRuntime runtime;
    runtime.initialize(1);  // Single worker to test preemption
    
    std::atomic<bool> heavy_task_running{false};
    std::atomic<bool> light_task_executed{false};
    
    // Heavy computation task
    runtime.kode([&heavy_task_running]() {
        heavy_task_running.store(true);
        std::cout << "[Heavy] Starting CPU-intensive task..." << std::endl;
        
        // Simulate heavy computation
        volatile long sum = 0;
        for (long i = 0; i < 100000000; i++) {
            sum += i;
            // Check for preemption periodically
            if (i % 10000000 == 0) {
                std::cout << "[Heavy] Progress: " << (i / 10000000) << "/10" << std::endl;
            }
        }
        
        std::cout << "[Heavy] Completed (sum: " << sum << ")" << std::endl;
        heavy_task_running.store(false);
    });
    
    // Light task may only run after the heavy task unless the heavy task yields.
    std::this_thread::sleep_for(std::chrono::milliseconds(50));  // Let heavy task start
    
    runtime.kode([&light_task_executed]() {
        std::cout << "[Light] Quick task executing!" << std::endl;
        light_task_executed.store(true);
    });
    
    runtime.join_all();
    
    if (light_task_executed.load()) {
        std::cout << "OK: Light task executed once the worker became available" << std::endl;
    } else {
        std::cout << "FAIL: Light task did not execute" << std::endl;
    }
    
    runtime.shutdown();
}

void test_timeout_functionality() {
    std::cout << "\n=== Test 4: Timeout Functionality ===" << std::endl;
    
    ConcurrencyRuntime runtime;
    runtime.initialize(2);
    
    // Task that should complete within timeout
    std::cout << "[Timeout] Testing task that completes in time..." << std::endl;
    runtime.with_timeout(std::chrono::milliseconds(200), []() {
        std::this_thread::sleep_for(std::chrono::milliseconds(100));
        std::cout << "[FastTask] Completed within timeout!" << std::endl;
    });
    
    std::this_thread::sleep_for(std::chrono::milliseconds(300));
    
    // Task that should timeout
    std::cout << "[Timeout] Testing task that times out..." << std::endl;
    runtime.with_timeout(std::chrono::milliseconds(100), []() {
        std::this_thread::sleep_for(std::chrono::milliseconds(300));
        std::cout << "[SlowTask] Work continued after cancellation request" << std::endl;
    });
    
    std::this_thread::sleep_for(std::chrono::milliseconds(200));
    
    runtime.join_all();
    runtime.shutdown();
}

void test_context_switching() {
    std::cout << "\n=== Test 5: Context Switching ===" << std::endl;
    
    ConcurrencyRuntime runtime;
    runtime.initialize(1);  // Single worker to force context switching
    
    std::atomic<int> switch_count{0};
    
    // Multiple tasks that yield frequently
    for (int i = 0; i < 3; i++) {
        runtime.kode([i, &switch_count, &runtime]() {
            for (int j = 0; j < 5; j++) {
                std::cout << "[Task " << i << "] Iteration " << j << std::endl;
                runtime.yield();  // Cooperative yield
                switch_count.fetch_add(1);
                std::this_thread::sleep_for(std::chrono::milliseconds(10));
            }
            std::cout << "[Task " << i << "] Completed" << std::endl;
        });
    }
    
    runtime.join_all();
    std::cout << "Total context switches: " << switch_count.load() << std::endl;
    runtime.shutdown();
}

void test_performance_benchmark() {
    std::cout << "\n=== Test 6: Performance Benchmark ===" << std::endl;
    
    const int NUM_TASKS = 1000;
    
    auto start_time = std::chrono::high_resolution_clock::now();
    
    ConcurrencyRuntime runtime;
    runtime.initialize();  // Auto-detect cores
    
    std::atomic<int> completed_tasks{0};
    
    // Spawn many lightweight tasks
    for (int i = 0; i < NUM_TASKS; i++) {
        runtime.kode([i, &completed_tasks]() {
            // Minimal work per task
            volatile int x = i * 2;
            (void)x;  // Suppress unused variable warning
            completed_tasks.fetch_add(1);
        });
    }
    
    runtime.join_all();
    
    auto end_time = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::microseconds>(end_time - start_time);
    
    std::cout << "Spawned and executed " << NUM_TASKS << " tasks in " 
              << duration.count() << " microseconds" << std::endl;
    std::cout << "Average time per task: " << (duration.count() / NUM_TASKS) << " microseconds" << std::endl;
    std::cout << "Tasks per second: " << (NUM_TASKS * 1000000 / duration.count()) << std::endl;
    
    runtime.shutdown();
}

int main() {
    std::cout << "=== Kode Concurrency System Tests ===" << std::endl;
    std::cout << "Testing Go-style goroutines, channels, and cooperative scheduling" << std::endl;
    
    try {
        test_basic_task_creation();
        test_channel_communication();
        test_preemptive_scheduling();
        test_timeout_functionality();
        test_context_switching();
        test_performance_benchmark();
        
        std::cout << "\n=== All Tests Completed ===" << std::endl;
        std::cout << "OK: Basic task creation and execution" << std::endl;
        std::cout << "OK: Channel-based communication (CSP)" << std::endl;
        std::cout << "OK: Cooperative scheduling" << std::endl;
        std::cout << "OK: Timeout and cancellation" << std::endl;
        std::cout << "OK: Context switching and yielding" << std::endl;
        std::cout << "OK: Performance benchmarking" << std::endl;
        
    } catch (const std::exception& e) {
        std::cerr << "Test failed with exception: " << e.what() << std::endl;
        return 1;
    }
    
    return 0;
}
